src/veranda/raster/native/geotiff.py

```diff
@@ -0,0 +1,498 @@
+
+            for band in self.bands:
+                self.src.GetRasterBand(band).Fill(int(self._nodatavals[band]))
+                self.src.GetRasterBand(band).SetNoDataValue(float(self._nodatavals[band]))
+                self.src.GetRasterBand(band).SetScale(self._scale_factors[band])
+                self.src.GetRasterBand(band).SetOffset(self._offsets[band])
+        else:
+            err_msg = f"Mode '{self.mode}' not known."
+            raise ValueError(err_msg)
+
+        if self.src is None:
+            err_msg = f"Open failed: {self.filepath}"
+            raise IOError(err_msg)
+
+    def read(self, row=0, col=0, n_rows=None, n_cols=None, bands=None, decoder=None, decoder_kwargs=None):
+        """
+        Read data from a GeoTIFF file.
+
+        Parameters
+        ----------
+        row : int, optional
+            Row number/index. Defaults to 0.
+        col : int, optional
+            Column number/index. Defaults to 0.
+        n_rows : int, optional
+            Number of rows of the reading window (counted from `row`). If None (default), then the full extent of
+            the raster is used.
+        n_cols : int, optional
+            Number of columns of the reading window (counted from `col`). If None (default), then the full extent of
+            the raster is used.
+        bands : int or tuple or list, optional
+            Band numbers of the GeoTIFF file to read data from. Defaults to None, i.e. all available bands will be
+            used.
+        decoder : callable, optional
+            Decoding function expecting a NumPy array as input.
+        decoder_kwargs : dict, optional
+            Keyword arguments for the decoder.
+
+        Returns
+        -------
+        data : dict
+            Dictionary mapping band numbers to NumPy arrays read from disk.
+
+        """
+
+        decoder_kwargs = decoder_kwargs or dict()
+        n_rows = self.shape[0] if n_rows is None else n_rows
+        n_cols = self.shape[1] if n_cols is None else n_cols
+        bands = bands or self.bands
+        bands = to_list(bands)
+
+        data = dict()
+        for band in bands:
+            band = int(band)
+            band_data = self.src.GetRasterBand(band).ReadAsArray(col, row, n_cols, n_rows)
+            scale_factor = self._scale_factors[band]
+            nodataval = self._nodatavals[band]
+            offset = self._offsets[band]
+            if self.auto_decode:
+                band_data = band_data.astype(float)
+                band_data[band_data == nodataval] = np.nan
+                band_data = band_data * scale_factor + offset
+            else:
+                if decoder is not None:
+                    dtype = GDAL_TO_NUMPY_DTYPE(self._dtypes[band])
+                    band_data = decoder(band_data, nodataval=nodataval, band=band, scale_factor=scale_factor,
+                                        offset=offset, dtype=dtype, **decoder_kwargs)
+            data[band] = band_data
```

Comment:
extract `{band: read_band(band)}`
Again I guess there is some code duplication

Reply:
Done.